### PR TITLE
Added start([phone|bot_token]) convenience method to quickly connect and authorize.

### PR DESCRIPTION
**Pull request by JosXa**

### CODE REVIEW

File: telethon/telegram_client.py [hunk 1]
```diff
@@ -184,6 +184,84 @@ def send_code_request(self, phone, force_sms=False):
...
+
+        if code_callback is None:
+            code_callback = lambda: input("Please enter the code you received: ")
+        else:
+            if not callable(code_callback):
+                raise ValueError("The code_callback parameter needs to be a callable function that "
+                                 "returns the code you received by Telegram.")
+
+        if password_callback is None:
+            password_callback = lambda: input("Please enter your password: ")
```
Comment (Lonami) [inline telethon/telegram_client.py]: We likely should use `getpass` instead.

File: telethon/telegram_client.py [hunk 2]
```diff
@@ -184,6 +184,84 @@ def send_code_request(self, phone, force_sms=False):
...
+        else:
+            if not callable(password_callback):
```
Comment (Lonami) [inline telethon/telegram_client.py]: Users probably know their 2FA password, and this won't let them pass the string directly.

File: telethon/telegram_client.py [hunk 3]
```diff
@@ -184,6 +184,84 @@ def send_code_request(self, phone, force_sms=False):
...
+            return self
+
+        if bot_token:
+            self(ImportBotAuthorizationRequest(
+                flags=0, bot_auth_token=bot_token,
+                api_id=self.api_id, api_hash=self.api_hash
+            ))
+        elif phone or self._phone:
+            phone = phone or self._phone
+            self.send_code_request(phone)
```
Comment (Lonami) [inline telethon/telegram_client.py]: Where are you using `force_sms`?

File: telethon/telegram_client.py [hunk 4]
```diff
@@ -184,6 +184,84 @@ def send_code_request(self, phone, force_sms=False):
...
+            provided_code = str(code_callback())
+            try:
+                # Normal login
+                self.sign_in(phone, provided_code)
+            except SessionPasswordNeededError:
+                # 2FA
+                provided_password = str(password_callback())
```
Comment (Lonami) [inline telethon/telegram_client.py]: If the user gave a string, use `getpass`. So we can retry while the password is wrong. Maybe raise as the user explicitly typed in a password and may not expect more wait for input?

File: telethon/telegram_client.py [hunk 5]
```diff
@@ -184,6 +184,84 @@ def send_code_request(self, phone, force_sms=False):
...
+                salt = self(GetPasswordRequest()).current_salt
+                self(CheckPasswordRequest(
```
Comment (Lonami) [inline telethon/telegram_client.py]: This doesn't take into consideration that the password may be wrong.

File: telethon/telegram_client.py [hunk 6]
```diff
@@ -184,6 +184,84 @@ def send_code_request(self, phone, force_sms=False):
...
+                flags=0, bot_auth_token=bot_token,
+                api_id=self.api_id, api_hash=self.api_hash
+            ))
+        elif phone or self._phone:
+            phone = phone or self._phone
+            self.send_code_request(phone)
+            provided_code = str(code_callback())
+            try:
+                # Normal login
+                self.sign_in(phone, provided_code)
```
Comment (Lonami) [inline telethon/telegram_client.py]: The code may be wrong, and it's not handled. Should re-ask the user for the code.

File: telethon/telegram_client.py [hunk 7]
```diff
@@ -1289,4 +1367,4 @@ def get_input_entity(self, peer):
             'Make sure you have encountered this peer before.'.format(peer)
         )
 
-    # endregion
+        # endregion
```
Comment (Lonami) [inline telethon/telegram_client.py]: I don't think this "automatic" indentation is right?

File: telethon/telegram_client.py (hunk 2 above)
Comment (JosXa) [reply telethon/telegram_client.py]: You're right. We had discussed that.

File: telethon/telegram_client.py (hunk 3 above)
Comment (JosXa) [reply telethon/telegram_client.py]: Fml

File: telethon/telegram_client.py [hunk 8]
```diff
@@ -239,25 +232,52 @@ def start(self, phone=None, bot_token=None, force_sms=False, code_callback=None,
...
+            attempts = 0
+            while attempts < 3:
+                try:
+                    # Normal login
+                    provided_code = str(code_callback())
+
+                    # raises SessionPasswordNeededError if 2FA enabled
+                    signed_in = bool(self.sign_in(phone, provided_code))
+                    if signed_in:
+                        print("Signed in successfully as {}".format(self.get_me()))
```
Comment (Lonami) [inline telethon/telegram_client.py]: `.get_me()` is a bit silly. `.sign_in` already returns the user, that's an unnecessary extra call…

File: telethon/telegram_client.py [hunk 9]
```diff
@@ -239,25 +232,52 @@ def start(self, phone=None, bot_token=None, force_sms=False, code_callback=None,
...
+                        break
+
+                except SessionPasswordNeededError:
+                    two_step_detected = True
+                    break
+                except (PhoneCodeEmptyError, PhoneCodeExpiredError,
+                        PhoneCodeHashEmptyError, PhoneCodeInvalidError) as e:
+                    print("Invalid code. Please try again...", file=sys.stderr)
```
Comment (Lonami) [inline telethon/telegram_client.py]: I don't get why people use ellipsis… For everything…

File: telethon/telegram_client.py [hunk 10]
```diff
@@ -239,25 +232,52 @@ def start(self, phone=None, bot_token=None, force_sms=False, code_callback=None,
...
+                    attempts += 1
+            else:
+                raise RuntimeError("Three consecutive sign-in attempts failed. Aborting...")
```
Comment (Lonami) [inline telethon/telegram_client.py]: Weird hardcoding "3" in two places.

File: telethon/telegram_client.py [hunk 11]
```diff
@@ -239,25 +232,52 @@ def start(self, phone=None, bot_token=None, force_sms=False, code_callback=None,
...
-                self.sign_in(phone, provided_code)
-            except SessionPasswordNeededError:
-                # 2FA
-                provided_password = str(password_callback())
-                salt = self(GetPasswordRequest()).current_salt
                 self(CheckPasswordRequest(
-                    helpers.get_password_hash(provided_password, salt)
+                    helpers.get_password_hash(str(password), salt)
                 ))
+            except Exception as e:
```
Comment (Lonami) [inline telethon/telegram_client.py]: This `except` is a bit too broad.

File: telethon/telegram_client.py [hunk 12]
```diff
@@ -239,25 +232,52 @@ def start(self, phone=None, bot_token=None, force_sms=False, code_callback=None,
...
+                raise ValueError("Invalid password for two-step verification.\n"
```
Comment (Lonami) [inline telethon/telegram_client.py]: Still not using the `from` syntax.

File: telethon/telegram_client.py [hunk 13]
```diff
@@ -315,8 +335,8 @@ def sign_in(self, phone=None, code=None,
                 result = self(SignInRequest(phone, phone_code_hash, code))
 
             except (PhoneCodeEmptyError, PhoneCodeExpiredError,
-                    PhoneCodeHashEmptyError, PhoneCodeInvalidError):
-                return None
+                    PhoneCodeHashEmptyError, PhoneCodeInvalidError) as e:
+                raise e  # Hand responsibility to the user
```
Comment (Lonami) [inline telethon/telegram_client.py]: `raise` is enough to re-raise. But why `except` it in the first place if we do no further processing?

**Pull request merged**